blog/2025-02-26-blockmax-wand/index.mdx

```diff
@@ -0,0 +1,228 @@
+
+[BM25](/blog/hybrid-search-explained#bm25) further refines TF-IDF by applying property length and frequency saturation normalization
+
+## Exhaustive Search 
+
+The exhaustive way of computing the [BM25](/blog/hybrid-search-explained#bm25) scores we use at Weaviate would be to check all the documents that have at least one of the query terms and score them. 
+
+But this is quite resource intensive; most searches are  for the top 10-100 results, and even with pagination, at most, only \-100 documents end up being shown to the user for each search. This means that if 100 000 documents have at least one of the query terms (very normal for queries with common words in databases with 1 million documents), this is 0.1% of the documents, many of which are completely irrelevant to the query, wasting a lot of CPU and I/O resources.
+
+## WAND
+
+[WAND (Weak AND)](https://dl.acm.org/doi/abs/10.1145/956863.956944) takes the inverted index and IDF to greatly reduce the number of documents we need to inspect when searching for the top-*k* documents that match a query.  
+It relies on two step search, to avoid ranking all documents for top-k search.
+
+* Approximate evaluation over query term postings in parallel to identify candidate docs with max impact heuristics (based on IDF);  
+* Promising candidates are fully evaluated, their exact scores are computed and they are added to the top-*k* results if the scores are higher than the lowest score.
+
+Max impact is the maximum score a term can contribute to the score.  
+Its upper bound is the IDF, e.g. a document with only the word *vector* will have a max impact equal to vector’s IDF. 
+
+* As we start to rank, we add enough documents to fill the top-*k* list;  
+* When we get *k* candidates, the list is full and we have a **lower bound** to beat, which is the score of the lowest ranked document;  
+* As we move forward, we can then start skipping documents where the sum of the max impacts of its terms is lower than the lower bound.
+
+WAND is what powers keyword search at Weaviate. The following section will show how effective it is.
+
+## BlockMax WAND
+
+While WAND works well and is already able to greatly reduce the number of documents that we need to inspect, it still has some limitations: it relies on a single global value of IDF for all documents in a term, which relies on the assumption that there may be a single document that just has that term.  
+[BlockMax WAND (BMW)](https://dl.acm.org/doi/10.1145/2009916.2010048) is WAND on steroids:
+* Divides posting lists into blocks with local max impact  
+* Skips and avoids decoding doc ids in blocks
+
+
+| WAND | BlockMax WAND |
+| :----: | :----: |
+| ![WAND](./img/wand_block.jpg) | ![BlockMax WAND](./img/blockmax_block.jpg) |
+
+BMW is a type of *meta*\-WAND, where we perform document skips at block level using max impact (shallow advances), and use the combination of max doc ids from blocks to even avoid loading whole blocks from disk.
+
```

Comment:
Wdyt about an intro for the table here, like:

This table shows an example posting, as output by BlockMAX WAND. You may notice that this includes some additional elements compared to postings for WAND shown above.